src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/SolutionExplorer_OutOfProc.cs

```diff
@@ -121,6 +121,13 @@ public void RenameFile(ProjectUtils.Project project, string oldFileName, string
             _instance.Workspace.WaitForAsyncOperations(FeatureAttribute.Workspace);
         }
 
+        public void RenameFileViaDTE(ProjectUtils.Project project, string oldFileName, string newFileName)
+        {
+            // Wireup to open files can happen asynchronously in the case we're being notified of changes on background threads.
+            _inProc.RenameFileViaDTE(project.Name, oldFileName, newFileName);
+            _instance.Workspace.WaitForAsyncOperations(FeatureAttribute.Workspace);
```

Comment:
We should add a TODO and tracking bug here becuase this *will* break once the project system makes further things async which they're currently working on.

Reply:
*This* code is correct (we only want to wait for workspace operations by default). Since the test itself makes an assertion on nav bars, what should it be waiting for before the assertion?

Reply:
This code is not correct, but cannot be made correct today. The project systems do (or will do soon in the case of csproj) design time builds asynchronously, and there is no strict guarantee that states that the design time build must be complete after the rename has been applied from the perspective of DTE. There might be code today that happens to force a synchronization but that's not in the contract. What is required *prior* to the WaitForAsyncOperation is a wait on the project system that it's pushed the work to the workspace, but the API to do didn't get completed in time for 16.0.

Reply:
➡️ Filed #33670 to handle this at a higher level when the APIs are available.